test: full-system integration test with Vortex

owickstrom opened the pull request:
Introducing _Vortex_, our new full-system integration test. This features a supervisor, which:

* starts a cluster of 3 replicas, running them as subprocesses
* starts the Vortex _workload_ as a separate process, which communicates with the cluster through a _driver_
* injects basic faults

The workload generates and submits commands and queries that check the consistency of results against a simple model.

Right now there's only a Zig (tb_client) driver available. The idea is that we can write drivers for each language client we support. Third-party clients might also be able to use Vortex to test their correctness.

Vortex can be run with a configurable duration. Our aim is to run it continuously on a VM later on. For now, it's integrated into CI as a smoke test. This can be run locally (Linux-only for now) with:

```shell
zig build test:integration -- "systest smoke"
```

You can also run it directly:

```shell
zig build vortex -Drelease && unshare -nfr zig-out/bin/vortex supervisor --tigerbeetle-executable=./tigerbeetle --test-duration-minutes=1
```

It uses [unshare(1)](https://man7.org/linux/man-pages/man1/unshare.1.html) to create a new namespace, with isolated network emulation and process tree. This enables us to run the supervisor without affecting the systems network overall, and we can run multiple instances of Vortex using the same (isolated) ports for replicas.

If you have permissions troubles with Ubuntu, see: https://github.com/YoYoGames/GameMaker-Bugs/issues/6015#issuecomment-2135552784

Future ideas for this work:

* Use unshare with a syscall in the supervisor, rather than through the shell
* More advanced fault injection (full partitions, packet reordering, filesystem corruption and read/write faults, clocks, etc)
* Upgrades (clients and replicas)
* Liveness checks
* Multiple drivers in a single test

owickstrom committed:
test: add basic systest supervisor

owickstrom committed:
test: run systest as integration smoke test

owickstrom commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    // let the replicas start
+    std.time.sleep(5 * std.time.ns_per_s);
```
Maybe we shouldn't?

owickstrom replied:
Doesn't do anything useful, I'm removing it.

matklad commented on .github/workflows/ci.yml:
```diff
@@ -48,6 +48,11 @@ jobs:
+        with:
+          distribution: "temurin"
+          java-version: "21"
+
```
Needing java to test our database is a rather significant regression. This is an intentional regression, so that we can iterate on systest infra quickly _and_ also get some coverage for the java client for free, but it's a regression non the less.

owickstrom it's on you to keep the tabs here and make sure we fix this eventually!

owickstrom replied:
Yeah, we should prioritize moving the workload to Zig and have a Zig driver as well (REPL or just direct client usage).

owickstrom replied:
Fixed!

matklad commented on build.zig:
```diff
@@ -641,8 +641,10 @@ fn build_test_integration(b: *std.Build, step_test_integration: *std.Build.Step,
+    integration_tests.root_module.addOptions("vsr_options", vsr_options);
     integration_tests.root_module.addOptions("test_options", integration_tests_options);
     const run_integration_tests = b.addRunArtifact(integration_tests);
+    run_integration_tests.setEnvironmentVariable("ZIG_EXE", b.graph.zig_exe);
```
Ditto here, ideally we shouldn't inject neither ver_options not ZIG_EXE here. Fine for now, but needs fixing

owickstrom replied:
Fixed!

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,245 @@
+const LoggedProcess = @import("./logged_process.zig");
+
+const assert = std.debug.assert;
+const log = @import("log.zig");
```
```suggestion
const log = std.log.scoped(.nemesis);
```

I _think_ we don't need the log.zig abstraction we are building here?

owickstrom replied:
You're right, it was mostly a convenience, but I'll scrap it.

matklad commented on src/testing/systest/supervisor/log.zig:
```diff
@@ -0,0 +1,63 @@
+    args: anytype,
+) void {
+    log_with_level(.info, scope, format, args);
+}
```
See comment in nemesis.zig --- I think we shouldn't be needing this abstraction.

owickstrom replied:
Ah, now I remember, it was because I wanted to use `std.log.scoped(...)` but couldn't make it work with the generated names replica1, replica2, etc. But we should make do without the fancy prefixed logs, it introduces too much extra cruft.

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,280 @@
+    return self.current_state.load(.seq_cst);
+}
+
+pub fn init(
```
init usually goes first

matklad replied:
and, as this is returning an allocated pointer, let's call these `create/destroy` rather than `inint/deinit`

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,280 @@
+) !*Self {
+    var arena = std.heap.ArenaAllocator.init(allocator);
+
+    const cwd = try std.process.getCwdAlloc(arena.allocator());
```
Memory leak here if process creating fails. You want to add

```zig
errdefer arena.deinit()
```

above

owickstrom replied:
Removing the arena entirely instead.

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,280 @@
+    self.expect_state_in(.{ .initial, .terminated, .completed });
+    defer self.expect_state_in(.{.running});
+
+    var child = std.process.Child.init(self.argv, self.allocator);
```
Same here --- any try below can fail, in which case this child process is getting leaked.

owickstrom replied:
I don't think `Child.init` heap-allocates? It's just a stack-allocated struct holding the passed-in allocator.

matklad replied:
after `.spawn`, child owns a file descriptor. This fd leaks if, eg, try format_argv fails (together with leaking the entire process)

owickstrom replied:
Ah, fixed now.

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,280 @@
+            }
+        }.log_stderr,
+        .{ child.stderr.?, self },
+    );
```
Let's avoid this thread for know and just inherit stderr. I am 0.8 sure that, given that we are at least somewhat cooperative, we could push the requirement to print unambiguous logs down to each child. And if we do do that, the system as a whole becomes significantly simpler.

Mabybe I am wrong, and we do need log wrapping, but I'd start with the assumption that we don't, and see how far we could push that

owickstrom replied:
Hmm, wouldn't concurrent logs jumble together if we don't use `std.log` (which has some lock around stderr)?

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,280 @@
+) !void {
+    self.expect_state_in(.{ .initial, .terminated, .completed });
+    defer self.expect_state_in(.{.running});
+
```
`assert(self.child == null)`

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,280 @@
+
+    const argv: []const []const u8 = &.{
+        "awk",
+        \\ BEGIN { for (i = 0; i < 10; i++) { print i > "/dev/fd/2"; } }
```
Let's remove `awk` dependency here

owickstrom replied:
Rewrote it with `yes`, is that better? I'm having a hard time finding something better that can be spawned for this test. Suggestions welcome!

owickstrom replied:
Fixed :+1:

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,245 @@
+const LoggedProcess = @import("./logged_process.zig");
+const Replica = @import("./replica.zig");
+const Nemesis = @import("./nemesis.zig");
+const log = @import("log.zig");
```
```suggestion
const log = std.log.scoped(.systest);
```

matklad commented on src/testing/systest/supervisor/replica.zig:
```diff
@@ -0,0 +1,5 @@
+
+name: []const u8,
+port: u16,
+process: *LoggedProcess,
```
Let's avoid adding a second `replica.zig` file, this is tiny enough to just inline at call-site

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,245 @@
+    random: std.rand.Random,
+    replicas: []const Replica,
+) !*Self {
+    const nemesis = try allocator.create(Self);
```
add `errdefer allocator.destroy(nemesis)`.

We don't have trys here yet, so this is dead code, but if we add a try below, the code would be incorrect. I guess you could also `errdefer comptime unreachable` if you really depend on anything below not being failable

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,245 @@
+netem_rules: NetemRules,
+
+pub fn init(
+    shell: *Shell,
```
Nemesis is a long-running event loop, so using shell (arena allocation) here is not right. 

Let's push as much as possible to static allocation here. For example, replica's slice should never be larger than, say, 16. That means that things like `matching` shouldn't be array lists and could be bounded arrays instead.

Similarly, I bet we could give a bound on the number of rules we need here, etc.

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,245 @@
+    }
+}
+
+fn random_replica_in_state(self: *Self, state: LoggedProcess.State) !?Replica {
```
There's no reason for this function to ever return an error

owickstrom replied:
Fixed

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,245 @@
+        return false;
+    }
+    try self.netem_rules.put(op, try self.allocator.dupe([]const u8, args));
+    assert(try self.netem_sync());
```
dupe above leaks if this try returns an erro

matklad replied:
I guess, either one dose not `try`, or one does not `alloc`. There's no `alloc`&`try` at the same time, that's too hard to get right!

matklad reviewed (commented):
Ok, so I am very satisfied with the big picture here, but I have doubts about the tactics-level implementation! I guess I have two kinds of concerns here:

* some things feel clearly wrong, but clearly temporary (us adding java dependency to our integration tests)
* but some things feel just wrong. For example, nemesis and systest are written in "let's allocate left and right and leak memory on errors". This is ok for provisional and temporary code, but these two things are going to be a core part of our very long-term testing strategy, and we should get this right. 
* and the meta concern here is that it's not super clear _which_ parts are "this is intentionally sloppy as we'll rewrite next week", and _which_ "this actually should be quite a bit more robust"

So, I think we have quite a bit to iterate on here still!

owickstrom commented:
OK, I think all easy fixes (sloppy memory management and such) are done now. Two things remaining essentially:

1. Log-wrapping of stderr, what should be done there
1. Rewriting workload in Zig
    1. Only using a Zig driver for the integration test in CI, and
    1. keeping the Java thing separate from our regular CI.
    1. This also, I believe, will remove the need for passing in ZIG_EXE and vsr_options into the integration test

matklad commented on src/integration_tests.zig:
```diff
@@ -484,3 +486,54 @@ test "in-place upgrade" {
+        try shell.exec("mvn clean install --batch-mode --quiet", .{});
+    }
+
+    log.info("Running 1m systest...", .{});
```
We _generally_ expect log messages to be prefixed, so we don't really start them with a capital letter.

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+
+pub fn main(shell: *Shell, allocator: std.mem.Allocator, args: CLIArgs) !void {
+    if (builtin.os.tag == .windows) {
+        log.err("supervisor: systest is not supported for Windows", .{});
```
```suggestion
        log.err("systest is not supported for Windows", .{});
```

We don't put general area prefixes here. We sometimes put the name of the function there.

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+    // Check that we are running as root
+    if (!std.mem.eql(u8, try shell.exec_stdout("id -u", .{}), "0")) {
+        log.err(
+            "This script needs to be run in a separate namespace using `unshare -nfr`",
```
```suggestion
            "this script needs to be run in a separate namespace using 'unshare -nfr'",
```

capitalization, and we conventionally use single quotes in log messages (not sure if that's the right convention, but that's, e.g, what our cli.zig uses)

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+    const tmp_dir = try shell.create_tmp_dir();
+    defer shell.cwd.deleteDir(tmp_dir) catch {};
+
+    // Check that we are running as root
```
```suggestion
    // Check that we are running as root.
```

Comments are proper sentences. This includes punctuation. 

https://github.com/tigerbeetle/tigerbeetle/blob/cd8f4e556c0cdb9726e838f21a73bc5a078fcb71/docs/TIGER_STYLE.md#L326-L327

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+    try shell.exec("ip link set up dev lo", .{});
+
+    log.info(
+        "supervisor: starting test with target runtime of {d}m",
```
```suggestion
        "starting test with target runtime of {d}m",
```

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+        const name = try shell.fmt("replica{d}", .{i});
+        const datafile = try shell.fmt("{s}/1_{d}.tigerbeetle", .{ tmp_dir, i });
+
+        // Format datafile
```
```suggestion
        // Format datafile.
```

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+const assert = std.debug.assert;
+
+const replica_count = 3;
+const replica_ports = [replica_count]u16{ 3000, 3001, 3002 };
```
maybe just

```
const cluster_address = "3000,3001,3002";
comptime {
    assert(std.mem.count(cluster_addresses, ",") == replica_count - 1)
}
```

?

That'll avoid leaky comma_separated_ports (it currently incorrectly handles allocation failures)

owickstrom replied:
Fixed `comma_separated_ports` instead. :) I have a strong suspicion we will want those port values. Maybe premature optimization now, but the fix made more sense to me.

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+        while (std.time.nanoTimestamp() - time_start < test_duration_ns) {
+            // Try to do something funky in the nemesis, and if it fails, wait for a while
+            // before trying again.
+            if (!try nemesis.wreak_havoc()) {
```
It's unclear what the boolean here means, and the comment doesn't really explain it. "fails" is  `try` part, but we also have a `bool`?

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+    // Let the workload finish by itself, or kill it after we've run for the required duration.
+    // Note that the nemesis is blocking in this loop.
+    const workload_result = term: {
+        while (std.time.nanoTimestamp() - time_start < test_duration_ns) {
```
I personally prefer writing these kinds of loops in the deadline form:

```zig
while (std.time.nanoTimestamp() < deadline)
```

With time, I always confuse if it is `a-b` or `b-a`, and the deadline form avoids that. 

Though, I am also known for _always_ messing up any retry-timeout related code, so don't listen to me on this one 🤣

owickstrom replied:
Let's go with that!

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+
+    // Let the workload finish by itself, or kill it after we've run for the required duration.
+    // Note that the nemesis is blocking in this loop.
+    const workload_result = term: {
```
don't need a block here, zig is python enough:

```
const result = while(...) {
   break x;
} else y;
```

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+        while (std.time.nanoTimestamp() - time_start < test_duration_ns) {
+            // Try to do something funky in the nemesis, and if it fails, wait for a while
+            // before trying again.
+            if (!try nemesis.wreak_havoc()) {
```
I am not sure if it makes sense to separate nemesis in this way. `nemesis` also manages the processes, so both this `main` function and `nemesis` are effectively vying for control over the replicas slice. So maybe factor it slightly differently?

Have a single thing that maintains the slice of replicas, the workload, and implements the event loop. Then, separate `netmem` in a separate file, which _doesn't_ have access to the rest of the state. 

So one file is all the state and event-loop, and the other is mostly stateless poking at OS interfaces to week havoc?

owickstrom replied:
OK, I have moved some responsibility from nemesis to the systest main control flow. Not the full slice of replicas and the termination action, though. It seems reasonable to keep these "bad" actions and their probabilities together in the replica, so I refrained from moving `terminate_replica` to systest.

owickstrom replied:
After sleeping on it, I'm gonna try moving more control to main. :)

owickstrom replied:
It's much simpler now. :+1:

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+
+    var process = try LoggedProcess.create(allocator, name, argv, .{ .env = &env });
+    try process.start();
+    return process;
```
I am not a fan of such two-step create/start split in the LoggerProcess API. Or rather, I am a fan of what you are getting at conceptually here, but I think this split is not quite right. The way I see it, we have two things here:

* a single instance of process with a specific pid
* a _program_ which we kill&restart, and which gives rise to many instances of processes

And LoggerProcess tires to model the second concept. But I think you actually can't model that cleanly, given our upgrade story. 

During upgrades, you might actually have _two_ tigerbeetle binaries, and you need to check both flows:

* replace the binary on disk, and check that the process notices that and re-execs
* kill the old process, start the new process using a new binary in the different location. 

So it feels like it is perhaps best to _not_ try to model this, and leave mapping of executables on disk to processes to the main event loop, and instead narrowly scope `LoggerProcess` to be a one-shot process/pid?

owickstrom replied:
OK, I've made LoggedProcess a one-shot process, and have the restart logic in a new Replica struct that maintains that lifecycle. That's where we can put the upgrade logic and such later. I'm also considering move the datafile formatting there, but I'm not sure.

Please let me know what you think about this new split.

matklad commented on src/testing/systest/supervisor/systest.zig:
```diff
@@ -0,0 +1,241 @@
+    switch (workload_result) {
+        .Exited => |code| {
+            if (code == 128 + std.posix.SIG.KILL) {
+                log.info("supervisor: workload terminated (SIGKILL) as requested", .{});
```
Hm, this is not right I think? 128+signal is the shell thing, but we don't have a shell here, so we have proper .Exited/.Signal thing here.

owickstrom replied:
When we used SIGTERM it needed to be handled this way, because the JVM exited with (128 + signal). See http://journal.thobe.org/2013/02/jvms-and-kill-signals.html.

But I think now that we use SIGKILL it might not be needed anymore. I'll try dropping it.

owickstrom replied:
Removed :+1:

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,329 @@
+//! to test fault tolerance.
+//!
+//! NOTE: some duplication exists between this and `tmp_tigerbeetle.zig` that could perhaps be
+//! unified.
```
As per comment in main, consider making this thing one-shot, and model restart by the caller creating a new logged_process

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,329 @@
+                    }
+                }
+            }
+        }.log_stderr,
```
Good point about stderr not having a global lock, but I still don't feel good about this code. I must confess, at this point, this is 80% gut feeling, and I am totally ready to eventually admit that I was wrong here, but just not right away :-) 

I think we need to do _something_ with logs and error streams, but I am very not sure that a separate stderr reading thread is the correct something _here_, and I fear that implementing a wrong solution would disincentivize us from finding the right, simple solution here. 

So, let's "do nothing" here, let's just inherit stderr stream, make a mental note to "fix this later", and revisit it later. 

I have a suspicion that this _might_ actually be fine, provided that the child processes obey three rules:

* Each process log is self-identifying 
* Each log message fits into 4k limit of kernel's atomic writes
* Each log message is a result of a single `.write` syscall

If all the above holds, then I thing the log would _not_ be messed up. And, given that we write the child processes, we have the power to make the above points true, and they _seems_ like good ideas nonetheless.

owickstrom replied:
Time for curveball: I think we should keep this, because I want to catch `err` logs from workload/replica and fail the systest on that. In the workload case, there shouldn't be error logs, just crashes, but for TB replicas we saw error logs without crashes. I'd rather be defensive here than relying on convention across all our future drivers and changes in replicas.

owickstrom replied:
And maybe there are reasonable cases where a replica should err-log but not crash?

matklad replied:
Hm, I'd maybe avoid parsing the logs. This is again, gut feeling, but for cfo, things definitely improved once we stopped trying to collect the logs, and instead went for binary "does it exit with zero". I think it makes more sense to treat this as a black box --- there's a certain contract that the cluster should fulfill, and it doesn't _really_ matter whether replica logs something or not.

owickstrom replied:
So, I tried, at it seems to hold so far. Let's give it a chance.

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,290 @@
+
+fn random_replica_in_state(self: *Self, state: LoggedProcess.State) ?*LoggedProcess {
+    var matching: [replicas_count_max]*LoggedProcess = undefined;
+    var pos: u8 = 0;
```
1. We don't generally contract words, so, `position` (curiously, I used to be a `let mut res` person before joining TB, but I am in the `var result` camp now!)
2. We don't usually use `position`, it is rather `index` or `count`

matklad commented on src/testing/systest/supervisor/nemesis.zig:
```diff
@@ -0,0 +1,290 @@
+        return null;
+    }
+    std.rand.shuffle(self.random, *LoggedProcess, matching[0..pos]);
+    return matching[0];
```
Why shuffle when you can select a random index? Alternatively, could go all the way to the reservoir sampling here (I _think_ Knuth shuffle is reservoir sampling in disguise anyway?)

owickstrom replied:
I actually don't know. Probably had some other idea at first and it got left over! Fixed.

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,299 @@
+
+// Passed in to init
+allocator: std.mem.Allocator,
+argv: []const []const u8,
```
Let's avoid storing argv. We aren't actually using it anywhere, and, if we don't store it, then the caller wouldn't have to dupe it!

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,299 @@
+    argv: []const []const u8,
+    options: Options,
+) !*Self {
+    const cwd = try std.process.getCwdAlloc(allocator);
```
Let's add a sys test-wide invariant that everything that runs, runs with systest-own cwd. That way, we could completely remove any mention of cwd from everywhere. 

It would be up to sestet's caller to place it in the correct directory.

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,299 @@
+    };
+
+    self.child.cwd = cwd;
+    self.child.env_map = options.env;
```
Same for the environment --- I think we can add a constraint that we don' use environmental variables and rely only on the explicit arguments. That way, we won't have to mess up with env. 

In general, we control all the processes that we've spawn, so we should totally add as much constraints on what the children can do to simplify the whole.

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,299 @@
+const Options = struct { env: ?*const std.process.EnvMap = null };
+
+// Passed in to init
+allocator: std.mem.Allocator,
```
Let's not store allocator here. Shell stores an allocator as an exception, but production code should take it as an explicit parameter in init/deinit. See

* https://zsf.zulipchat.com/#narrow/stream/454371-std/topic/Xunmanaged.20-.3E.20X
* https://matklad.github.io/2020/12/28/csdi.html

for some background info on the pattern

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,299 @@
+            break :kill std.posix.kill(self.child.id, std.posix.SIG.KILL);
+        }
+    } catch |err| {
+        std.debug.print(
```
log.err

In general, never use `std.debug.print` (I used it a bunch in various auxiliary support files, and I now regret that 🤣 )

matklad commented on src/testing/systest/supervisor/logged_process.zig:
```diff
@@ -0,0 +1,299 @@
+    unreachable;
+}
+
+fn format_argv(allocator: std.mem.Allocator, argv: []const []const u8) ![]const u8 {
```
Feels like its a dead code now?

matklad commented on src/testing/systest/supervisor/network_faults.zig:
```diff
@@ -0,0 +1,173 @@
+        ));
+    }
+
+    // Everything stack-allocated and simple up to here. Now this feels like a shame just for
```
It feels indeed! Would be great if this were a thing:

```zig
log.info("syncing netem: {}", .{stdx.fmt_join("{}", " ", args.items)});

// in stdx.zig
fn fmt_join(comptime item_fmt: []const u8, sep: []const u8, items: anytype)
```

I _think_ this should be possible, although I don't remember the relevant fmt APIs off the top of my head my head. Could you them up and implement this(maybe as a separate PR to stdx even)?  I remember wanting this a couple of time and chickening out of actually doing the work :-)

matklad replied:
https://github.com/ziglang/zig/blob/ea527f7a850f0200681630d8f36131eca31ef48b/lib/std/fmt.zig#L833-L840 is the pattern to follow

owickstrom committed:
test: port basic workload and driver to zig

matklad commented on src/testing/systest/driver.zig:
```diff
@@ -0,0 +1,147 @@
+    try writer.writeInt(u32, @intCast(events.len), .little);
+
+    const bytes = std.mem.sliceAsBytes(events);
+    const size_written = try writer.write(bytes);
```
This wants to be a `writeAll` I think?

owickstrom replied:
Good catch! That API difference is subtle...

owickstrom commented:
> OK, I think all easy fixes (sloppy memory management and such) are done now. Two things remaining essentially:
>
> *  Log-wrapping of stderr, what should be done there
>  *  Rewriting workload in Zig 

Both fixed now.

Things I'll do in a separate PR:

* Write a Java driver, so that we can remove the old systest workload
* Implement retries in the Zig workload

matklad commented on build.zig:
```diff
@@ -622,16 +633,30 @@ fn build_test_integration(b: *std.Build, step_test_integration: *std.Build.Step,
+        .vsr_options = vsr_options,
+        .target = options.target,
+        .mode = options.mode,
+    });
```
Let's maybe pass vortex lazy_path through `options`? Right now, the `build voretx` and the vortex used by integration tests are different, so you'll get unnecessary rebuild.

The fact that integration tests build a separate copy of tiger beetle is an exception --- we intentionally use different option from the main tiger beetle build. Here, we could re-use vortex as is.

owickstrom replied:
I think we can't, actually, because the versions from `vsr_options` have to match, otherwise the client can't talk to the replicas (because it would be on version 0.0.1).

owickstrom replied:
Closing this, feel free to reopen.

matklad replied:
Yeah, you are right, I missed that we use a different version here!

matklad commented on build.zig:
```diff
@@ -622,16 +633,30 @@ fn build_test_integration(b: *std.Build, step_test_integration: *std.Build.Step,
         .filters = b.args orelse &.{},
     });
     integration_tests.root_module.addOptions("test_options", integration_tests_options);
+    integration_tests.step.dependOn(step_vortex);
```
You shouldn't need an explicit dependency. The thing about the `LazyPath` is that it auto-magically tracks deps. So, just `integration_tests_options.addOptionPath("vortex_exe", vortex_exe);` already adds the required dep!

owickstrom replied:
Nice, trying that.

owickstrom replied:
Sadly didn't work.

owickstrom replied:
Fixed!

matklad commented on build.zig:
```diff
@@ -798,6 +823,33 @@ fn build_scripts(
+    });
+
+    vortex.root_module.addOptions("vsr_options", options.vsr_options);
+    vortex.linkLibC();
```
Why do we need `linkLibC`? libc is not nodejs-level banned, but is definitely not encouraged, a kernel is all you need ;)

matklad replied:
Ahhh, that is because we use c client in the default driver! Smart! And we need libc for the C client for a bad reason (current version of Zig requires it for cImport).

owickstrom replied:
That's it!

matklad commented on src/testing/vortex/supervisor.zig:
```diff
@@ -0,0 +1,336 @@
+
+// Create a new Vortex-specific temporary directory. Caller owns the returned memory.
+fn create_tmp_dir(allocator: std.mem.Allocator) ![]const u8 {
+    const root = process.getEnvVarOwned(allocator, "TMPDIR") catch try allocator.dupe(u8, "/tmp");
```
Let's use `.zig-cache/tmp` for our temp dir. I think we can assume that supervisor is being run with cwd pointing at the repository root? 

That way, we don't depend on ambient environment (TMPDIR) and actually keep stuff scoped to the project. And other things already use zig-cache/tmp, so we also gain uniformity here.

owickstrom replied:
It feels a little off. I don't have super strong opinions about it but it would preclude us from:

1. compiling Vortex, shipping the executable to some VM, running it there (might not be a use case, ever)
2. lettings others use Vortex outside or repo (this seems like a better reason to make it cwd-portable)

owickstrom replied:
We could do a fallback solution? Check if `.zig-cache/tmp` exists and use that, or `TMPDIR`, or `/tmp`. But it's feels like bad magic.

matklad replied:
I think if someone uses vortex outside of the repo, they could live with `.zig-cache` created in the cwd?

owickstrom replied:
I guess? It's a little weird but probably not a problem.

owickstrom replied:
Fixed.

matklad commented on src/testing/vortex/supervisor.zig:
```diff
@@ -0,0 +1,336 @@
+    defer allocator.free(tmp_dir);
+    defer std.fs.cwd().deleteTree(tmp_dir) catch |err| {
+        log.err("failed deleting temporary directory ({s}): {any}", .{ tmp_dir, err });
+    };
```
Let's do a single

```
defer {

};
```

here to avoid thinking about the order of evaluation.

matklad commented on src/testing/vortex/supervisor.zig:
```diff
@@ -0,0 +1,336 @@
+
+    var replicas: [replica_count]*Replica = undefined;
+    inline for (0..replica_count) |i| {
+        var datafile_buf: [1024]u8 = undefined;
```
Use `std.fs.max_path_bytes` here

matklad commented on src/testing/vortex/supervisor.zig:
```diff
@@ -0,0 +1,336 @@
+            datafile,
+        } });
+        defer allocator.free(result.stderr);
+        defer allocator.free(result.stdout);
```
Same here, merge defers. Defer can not fail, so it is always correct to merge them.

matklad commented on src/testing/vortex/supervisor.zig:
```diff
@@ -0,0 +1,336 @@
+                log.err("failed formatting datafile: {s}", .{result.stderr});
+                process.exit(1);
+            },
+        }
```
I think you can do

```zig
        if (std.meta.eql(result.term, .{ .Exited = 0})) {
            
        }
```

here?

matklad commented on src/testing/vortex/supervisor.zig:
```diff
@@ -0,0 +1,336 @@
+
+    var replicas: [replica_count]*Replica = undefined;
+    inline for (0..replica_count) |i| {
+        var datafile_buf: [1024]u8 = undefined;
```
Missing 

```
        errdefer for (replicas[0..i]) |replica| replica.destroy();
```

to free stuff if the middle of the loop blows up

matklad replied:
also maybe `s/i/replica_index`?

owickstrom replied:
Had a `defer` for replicas below that I'm moving up.

matklad commented on src/testing/vortex/logged_process.zig:
```diff
@@ -0,0 +1,259 @@
+
+pub fn terminate(
+    self: *Self,
+    signal: comptime_int,
```
Why pass the signal here? We should only ever support abrupt, ungraceful shutdown, so this always will be a sigkill.

owickstrom replied:
Leftover, I did an experiment with SIGTERM that didn't pan out. Removing again.

matklad commented on src/testing/vortex/replica.zig:
let's avoid the second `replica.zig` file in the codebase. `replica.zig` is the main file of the entire thing, it is valuable that you can unambiguously go to definition to it.

owickstrom replied:
Just rename, or do you think I should put it in `supervisor.zig`?

owickstrom replied:
It's only about 100 LOC.

owickstrom replied:
Done!

matklad commented on src/testing/vortex/replica.zig:
```diff
@@ -0,0 +1,96 @@
+            .completed => return State.completed,
+        }
+    } else return State.initial;
+}
```
```suggestion
    if (self.process) |process| {
        switch (process.state()) {
            .running => return .running,
            .terminated => return .terminated,
            .completed => return .completed,
        }
    } else return .initial;
}
```

matklad commented on src/testing/vortex/replica.zig:
```diff
@@ -0,0 +1,96 @@
+        "--addresses={s}",
+        .{self.replica_addresses},
+    );
+    defer self.allocator.free(addresses);
```
let's pre-compute this in init to avoid a random allocation here.

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,407 @@
+    @"driver-command": []const u8,
+};
+
+pub fn main(allocator: std.mem.Allocator, args: CLIArgs) !void {
```
Let's make the interface more crisp here. Instead of workload spawning the driver, let's just pass in driver-s pipes:

```
fn main(
  allocator, 
  args, 
  pipes: struct {in : std.fs.File, out: std.fs.File }

)
```

and let the superviser to spawn the driver.

owickstrom replied:
I'm not sure what you mean here. The workload runs in a separate process, so we'd have to pass in file descriptors? Or am I missing something?

owickstrom replied:
e.g. `/proc/<PID>/fd/0`, with the PID of the externally spawned driver?

owickstrom replied:
OK, interface is now a lot crisper, and that let me delete a good amount of code as well!

owickstrom committed:
review: address review comments

* simplify exit code check
* move defers up to cover failed initialization
* remove explicit enum qualifier
* precompute replica addresses arg
* index naming fix
* fix integration test dependency to exe paths
* use .zig-cache as temp dir
* move replica into supervisor.zig
* lift driver spawning up to vortex entry point

owickstrom committed:
test: write results from main thread in Zig driver

Instead of writing to stdout from the tb_client callback directly, we
use a mutex and a condition to await the results in the main thread,
through a shared context.

One reason for doing this is that we can reinitialize the client in
a controlled way. Previously, a callback might happen after
reinitializing the client.

This is based on the same mechanism as used in
`src/clients/c/samples/main.zig`.

Also:
* handle stdout broken pipe in zig driver
* pass in buffer of known size with correct lifetime

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+//! This workload runs in an loop, generating and executing operations on a cluster through a
+//! _driver_.
+//!
+//! Any sucessful operations are reconciled with a model, tracking what accounts exist. Future
```
```suggestion
//! Any successful operations are reconciled with a model, tracking what accounts exist. Future
```

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+    allocator: std.mem.Allocator,
+    driver: *const DriverStdio,
+) !void {
+    var model = Model{
```
It's OK to use dynamic allocation here for now, but I'd seriously consider refactoring
this to static alloc, to make code _easier_ to understand. Static alloc is harder to code for
the first time, as you need to actually _understand_ how many accounts/pending transfers you
could have, but, once understood and encoded in the source code, this understanding can be
re-used for free by future readers.

In terms of code, that's s/AutoHashMap/AutoHashMapUnmanaged

owickstrom replied:
Makes sense. I did look around for how to do that, but might need some guidance. Let's revisit in a future PR!

owickstrom replied:
Fixed now instead. :)

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+    }
+}
+
+var accounts_buffer = std.mem.zeroes([accounts_count_max]tb.Account);
```
Normally, I would advice against using global variables in such a way, but it seems like the workload is a genuine exception, where the globals make sense!

Though, we _do_ pass model as a parameter explicitly everywhere, so maybe just pull this over to model?

owickstrom replied:
I see your point, but model would be a strange place to keep it. It's really only a concern for the workload itself.

owickstrom replied:
I can, however, lift the `accounts_buffer` into `main`.

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+var result_buffers: ResultBuffers = std.mem.zeroes(ResultBuffers);
+
+fn execute(command: Command, driver: *const DriverStdio) !?Result {
+    switch (command) {
```
Could this be `inline else => |tag|` to avoid code duplication?

owickstrom replied:
Not quite, because the `Command` tag enum isn't the same as `StateMachine.Operation`. I think we'll need to live with the somewhat duplicated code, because it would create a weird coupling if we tried to unify them.

matklad replied:
Hm, the following works, but, yeah, its a bit on the ugly side:

```zig
fn execute(command: Command, driver: *const DriverStdio) !?Result {
    switch (command) {
        inline else => |events, tag| return execute_regular(
            tag,
            if (tag == .lookup_all_accounts)
                .lookup_accounts
            else
                std.enums.nameCast(StateMachine.Operation, tag),
            events,
            driver,
        ),
    }
}
```

owickstrom replied:
I'll see if I can clean it up.

owickstrom replied:
Fixed! Found a nicer way to slice it, avoiding the `execute_regular` helper.

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+                accounts_created[0..accounts_new.len],
+            ) |account_new, *account_created| {
+                account_created[0] = .ok;
+                account_created[1] = account_new;
```
Feels like a useless copy here? Can you use accounts_new as is below? That is,
have two parallels arrays rather than an array of pairs?

owickstrom replied:
Fixed!

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+                if (transfer.flags.pending) {
+                    try testing.expect(!model.pending_transfers.contains(transfer.id));
+                    try model.pending_transfers.put(transfer.id, {});
+                    // We can't let this grow unboundedly:
```
Ah, so if we assert this here, there's no reason _not_ to use putAssumingCapacity!

owickstrom replied:
OK, fixed!

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+    pending_transfers: std.AutoHashMap(u128, void),
+
+    // O(n) lookup, but it's limited by `accounts_count_max`, so it's OK for this test.
+    fn account_exists(self: @This(), id: u128) bool {
```
Maybe model rather than self? We try to use self less, to keep the naming unfimorm between methods and functions.

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+    }
+}
+
+fn random_create_accounts(random: std.Random, model: *const Model) !Command {
```
No need for `!` here?

matklad commented on src/testing/vortex/workload.zig:
```diff
@@ -0,0 +1,432 @@
+    return .{ .create_accounts = events[0..events_count] };
+}
+
+fn random_create_transfers(random: std.Random, model: *const Model) !Command {
```
and here?

matklad commented on src/testing/vortex/zig_driver.zig:
```diff
@@ -0,0 +1,128 @@
+                try writer.writeInt(u32, @intCast(count), .little);
+                try writer.writeAll(result);
+            } else {
+                log.err(
```
Do we want to crash here? Otherwise, the workload might hang I think?

owickstrom replied:
Done.

matklad commented on src/testing/vortex/zig_driver.zig:
```diff
@@ -0,0 +1,128 @@
+            const buffer_size = @sizeOf(StateMachine.Event(comptime_operation)) * count;
+            assert(try reader.readAtLeast(buffer[0..buffer_size], buffer_size) == buffer_size);
+
+            return .{ comptime_operation, buffer[0..buffer_size] };
```
Variable `buffer` does not live long enough ;-)

owickstrom replied:
Fixing this in https://github.com/tigerbeetle/tigerbeetle/pull/2422. Merge into this PR first, or into main afterwards?

owickstrom replied:
Merged into this PR now.

matklad replied:
Looks like it still returns a slice pointing to stack-allocated data, no?

owickstrom replied:
Argh, not sure what happened here. Fixing tomorrow.

owickstrom replied:
Fixed.

owickstrom committed:
review: address second round of feedback

* lift account_buffers into main
* avoid copying and array of tuples
* use unmanaged hashmap for pending transfers
* explicitly name model
* remove unnecessary error types
* panic zig driver on incorrect message size
* remove debug log
* clean up operation/command tag mapping

Co-authored-by: Alex Kladov <aleksey.kladov@gmail.com>

owickstrom committed:
test: use explicit flag names for unshare

owickstrom committed:
review: fix lifetime issue by lifting buffers up
